docs/src/developers/server/config.md
# Configuration

The config of the FacilMap server can be set either by using environment variables (useful for docker) or by editing `config.env`.

| Variable              | Required | Meaning                                                                                                                          |
|-----------------------|----------|----------------------------------------------------------------------------------------------------------------------------------|
| `USER_AGENT`          | *        | Will be used for all HTTP requests (search, routing, GPX/KML/OSM/GeoJSON files). You better provide your e-mail address in here. |
| `APP_NAME`            |          | If specified, will replace “FacilMap” as the name of the app throughout the UI. |
| `TRUST_PROXY`         |          | Whether to trust the X-Forwarded-* headers. Can be `true` or a comma-separated list of IP subnets (see the [express documentation](https://expressjs.com/en/guide/behind-proxies.html)). Currently only used to calculate the base URL for the `opensearch.xml` file. |
| `BASE_URL`            |          | If `TRUST_PROXY` does not work for your particular setup, you can manually specify the base URL where FacilMap can be publicly reached here. |
| `HOST`                |          | The ip address to listen on (leave empty to listen on all addresses) |
| `PORT`                |          | The port to listen on.<br><br>*Default:* `8080` |
| `DB_TYPE`             |          | The type of database. Either `mysql`, `postgres`, `mariadb`, `sqlite`, or `mssql`.<br><br>*Default:* `mysql` |
| `DB_HOST`             |          | The host name of the database server.<br><br>*Default:* `localhost` |
| `DB_PORT`             |          | The port of the database server (optional). |
| `DB_NAME`             |          | The name of the database.<br><br>*Default:* `facilmap` |
| `DB_USER`             |          | The username to connect to the database with.<br><br>*Default:* `facilmap` |
| `DB_PASSWORD`         |          | The password to connect to the database with.<br><br>*Default:* `facilmap` |
| `ORS_TOKEN`           |          | [OpenRouteService API key](https://openrouteservice.org/). If not specified, advanced routing settings will not be shown. |
| `MAPBOX_TOKEN`        |          | [Mapbox API key](https://www.mapbox.com/signup/). If neither this nor `ORS_TOKEN` are specified, the routing tab and any routing options will be hidden. |
| `MAXMIND_USER_ID`     |          | [MaxMind user ID](https://www.maxmind.com/en/geolite2/signup). |
| `MAXMIND_LICENSE_KEY` |          | MaxMind license key. |
| `LIMA_LABS_TOKEN`     |          | [Lima Labs](https://maps.lima-labs.com/) API key (for Lima Labs map style) |
| `THUNDERFOREST_TOKEN` |          | [Thunderforest](https://www.thunderforest.com/) API key (for OpenCycleMap map style) |
| `TRACESTRACK_TOKEN`   |          | [Tracestrack](https://tracestrack.com/) API key (for Tracestrack Topo map style) |
| `HIDE_COMMERCIAL_MAP_LINKS` |    | Set to `1` to hide the links to Google/Bing Maps in the “Map style” menu. |
| `CUSTOM_CSS_FILE`     |          | The path of a CSS file that should be included ([see more details below](#custom-css-file)). |
| `NOMINATIM_URL`       |          | The URL to the Nominatim server (used to search for places).<br><br>*Default:* `https://nominatim.openstreetmap.org` |
| `OPEN_ELEVATION_URL`  |          | The URL to the Open Elevation server (used to look up the elevation for markers).<br><br>*Default:* `https://api.open-elevation.com` |
| `OPEN_ELEVATION_THROTTLE_MS` |   | The minimum time between two requests to the Open Elevation API. Set to `0` if you are using your own self-hosted instance of Open Elevation.<br><br>*Default:* `1000` |
| `OPEN_ELEVATION_MAX_BATCH_SIZE` | | The maximum number of points to resolve in one request through the Open Elevation API. Set this to `1000` if you are using your own self-hosted Open Elevation instance.<br><br>*Default:* `200` |
| `DONATE_URL`          |          | Define a custom target for the “Donate” button. If you decide to link your own donation page to cover the costs of your hosting, consider mentioning FacilMap’s donation page there for the costs of the development of the software. You can also set this to an empty string to completely hide the donation button.<br><br>*Default:* `https://docs.facilmap.org/users/contribute/` |

FacilMap makes use of several third-party services that require you to register (for free) and generate an API key:
* Mapbox and OpenRouteService are used for calculating routes. Mapbox is used for basic routes, OpenRouteService is used when custom route mode settings are made. If these API keys are not defined, calculating routes will fail.
* Maxmind provides a free database that maps IP addresses to approximate locations. FacilMap downloads this database to decide the initial map view for users (IP addresses are looked up in FacilMap’s copy of the database, on IP addresses are sent to Maxmind). This API key is optional, if it is not set, the default view will be the whole world.
* Lima Labs is used for nicer and higher resolution map tiles than Mapnik. The API key is optional, if it is not set, Mapnik will be the default map style instead.

## Custom CSS file

To include a custom CSS file in the UI, set the `CUSTOM_CSS_FILE` environment variable to the file path.

When running FacilMap with docker, you can mount your CSS file as a volume into the container, for example with the following docker-compose configuration:
```yaml
		environment:
			CUSTOM_CSS_FILE: /opt/facilmap/custom.css
		volumes:
			- ./custom.css:/opt/facilmap/custom.css
```

Your custom CSS file will be included in the map UI and in the table export. You can distinguish between the two by using the `html.fm-facilmap-map` and `html.fm-facilmap-table` selectors.

server/src/config.ts
import type { Options as SequelizeOptions } from "sequelize";
import "dotenv/config";
import { setConfig, setFetchAdapter } from "facilmap-utils";
import { configure } from "osm-api";

export interface DbConfig {
	type: SequelizeOptions['dialect'];
	host: SequelizeOptions['host'];
	port: SequelizeOptions['port'];
	database: string;
	user: string;
	password: string;
}

export interface Config {
	appName: string;
	userAgent: string;
	trustProxy?: boolean | string | string[] | number | ((ip: string) => boolean);
	baseUrl?: string;
	host?: string;
	port: number;
	db: DbConfig;
	orsToken?: string;
	mapboxToken?: string;
	maxmindUserId?: string;
	maxmindLicenseKey?: string;
	limaLabsToken?: string;
	thunderforestToken?: string;
	tracestrackToken?: string;
	/** Hide the "Open this on Google/Bing Maps" links in the map style menu */
	hideCommercialMapLinks?: boolean;
	customCssFile?: string;
	nominatimUrl: string;
	openElevationApiUrl: string;
	openElevationThrottleMs: number;
	openElevationMaxBatchSize: number;
	donateUrl: string;
}

const config: Config = {
	appName: process.env.APP_NAME || "FacilMap",
	userAgent: process.env.USER_AGENT || 'FacilMap',
	trustProxy: (
		!process.env.TRUST_PROXY ? undefined :
		process.env.TRUST_PROXY === "true" ? true :
		process.env.TRUST_PROXY.match(/^\d+$/) ? Number(process.env.TRUST_PROXY) :
		process.env.TRUST_PROXY
	),
	baseUrl: process.env.BASE_URL ? (process.env.BASE_URL.endsWith("/") ? process.env.BASE_URL : `${process.env.BASE_URL}/`) : undefined,
	host: process.env.HOST || undefined,
	port: process.env.PORT ? Number(process.env.PORT) : 8080,
	db : {
		type: process.env.DB_TYPE as any || "mysql", // mysql, postgres, mariadb, sqlite
		host: process.env.DB_HOST || "localhost",
		port: Number(process.env.DB_PORT) || undefined,
		database: process.env.DB_NAME || "facilmap",
		user: process.env.DB_USER || "facilmap",
		password: process.env.DB_PASSWORD || "facilmap"
	},
	orsToken: process.env.ORS_TOKEN || "", // Get a token on https://go.openrouteservice.org/
	mapboxToken: process.env.MAPBOX_TOKEN || "", // Get an API key on https://www.mapbox.com/signup/

	// Maxmind configuration. If specified, the maxmind GeoLite2 database will be downloaded for Geo IP lookup (to show the initial map state) and kept it in memory
	// Sign up here: https://www.maxmind.com/en/geolite2/signup
	maxmindUserId: process.env.MAXMIND_USER_ID || "",
	maxmindLicenseKey: process.env.MAXMIND_LICENSE_KEY || "",

	limaLabsToken: process.env.LIMA_LABS_TOKEN || "", // Get a token on https://maps.lima-labs.com/
	thunderforestToken: process.env.THUNDERFOREST_TOKEN || "", // Get a token on https://www.thunderforest.com/
	tracestrackToken: process.env.TRACESTRACK_TOKEN || "", // Get a token on https://tracestrack.com/

	hideCommercialMapLinks: process.env.HIDE_COMMERCIAL_MAP_LINKS === "1",

	customCssFile: process.env.CUSTOM_CSS_FILE || undefined,

	nominatimUrl: process.env.NOMINATIM_URL || "https://nominatim.openstreetmap.org",

	openElevationApiUrl: process.env.OPEN_ELEVATION_URL || "https://api.open-elevation.com",
	openElevationThrottleMs: process.env.OPEN_ELEVATION_THROTTLE_MS ? Number(process.env.OPEN_ELEVATION_THROTTLE_MS) : 1000, // Maximum one request per second, see https://github.com/Jorl17/open-elevation/issues/3
	openElevationMaxBatchSize: process.env.OPEN_ELEVATION_MAX_BATCH_SIZE ? Number(process.env.OPEN_ELEVATION_MAX_BATCH_SIZE) : 200,

	donateUrl: process.env.DONATE_URL ?? "https://docs.facilmap.org/users/contribute/", // Empty string means don't show donation button
};

setConfig({
	openElevationApiUrl: config.openElevationApiUrl,
	openElevationThrottleMs: config.openElevationThrottleMs,
	openElevationMaxBatchSize: config.openElevationMaxBatchSize,
	nominatimUrl: config.nominatimUrl
});

setFetchAdapter(async (input, init) => {
	const headers = new Headers(init?.headers);
	headers.set("User-Agent", config.userAgent);
	return await fetch(input, { ...init, headers });
});

configure({
	userAgent: config.userAgent
});

export default config;
